pil-analyzer/src/type_inference.rs

```diff
@@ -0,0 +1,675 @@
+                                .unifier
+                                .unify_types(t.clone(), Type::Fe)
+                                .map_err(|err| {
+                                    format!(
+                                        "Return type is expected to be `fe`, but got:{}.\n{err}",
+                                        self.substitute_to(t)
+                                    )
+                                }),
+                        }
+                    }
+                    Type::Array(ArrayType {
+                        base,
+                        length: Some(_),
+                    }) if base.as_ref() == &Type::Expr => {
+                        // An array of intermediate columns with fixed length. We ignore the length.
+                        // The condenser will have to check the actual length.
+                        let arr = Type::Array(ArrayType {
+                            base: base.clone(),
+                            length: None,
+                        });
+                        self.expect_type(&arr, value).map_err(|e| {
+                            format!("Expected dynamically-sized array for symbol {name}:\n{e}")
+                        })
+                    }
+                    t => self.expect_type(t, value),
+                }
+            } else {
+                self.process_expression(value).map(|ty| {
+                    self.inferred_types.insert(name.to_string(), ty);
+                })
+            })
+            .map_err(|e| format!("Error type checking the symbol {name} = {value}:\n{e}",))?;
+        }
+
+        self.check_expressions(expressions)?;
+
+        // From this point on, the substitutions are fixed.
+
+        // Now we check for all symbols that are not declared as a type scheme that they
+        // can resolve to a concrete type.
+        for (name, declared_type) in &self.declared_types {
+            if declared_type.vars.is_empty() {
+                // It is not a type scheme, see if we were able to derive a concrete type.
+                let inferred = self.substitute_to(declared_type.ty.clone());
+                if !inferred.is_concrete_type() {
+                    let inferred_scheme = self.to_type_scheme(inferred);
+                    Err(format!(
+                        "Could not derive a concrete type for symbol {name}.\nInferred type scheme: {}\n",
+                        format_type_scheme_around_name(
+                            name,
+                            &Some(inferred_scheme),
+                        )
+                    ))?;
+                }
+            }
+        }
+
+        // We have to check type schemes last, because only at this point do we know
+        // that other types that should be concrete do not occur as type variables in the
+        // inferred type scheme any more.
+        let type_var_mapping = self.inferred_types.iter().map(|(name, inferred_type)| {
```

Comment:
It seems `inferred_types` doesn't need to be a member of `Self`, as it's only used inside this function. As a variable, you could `into_iter` here

Reply:
Oh wow, great observation! Will change!